### PR TITLE
[MRG] Rolling hash, attempt N

### PR DESCRIPTION
**Pull request by betatim**
This is a new attempt at adding a rolling hash.

This uses the cyclic hash from lemire/rollinghash. The code is mostly a copy&paste from that repository, slightly modified to use the c++ stdlib RNG and allow seeding.

Design question: should we add a new `Counttable` like class that uses this or (as done for tests here) replace the hash used by `Counttable`?

Another thing I noticed: it is pretty tedious to test that this works properly (ended up adding print statements...) is it worth exposing these kmer iterators to python land? Other ideas to test that this does what it should be doing? In particular checking that the logic in the iterator is right.

- [ ] Is it mergeable?
- [ ] `make test` Did it pass the tests?
- [ ] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [ ] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [ ] Did it change the command-line interface? Only backwards-compatible
  additions are allowed without a major version increment. Changing file
  formats also requires a major version number increment.
- [ ] For substantial changes or changes to the command-line interface, is it
  documented in `CHANGELOG.md`? See [keepachangelog](http://keepachangelog.com/)
  for more details.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?
- [ ] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)

### CODE REVIEW

**Comment by betatim** [comment]
Previous PRs on this topic: #1622 #1699

**Comment by standage** [comment]
> Design question: should we add a new `Counttable` like class that uses this or (as done for tests here) replace the hash used by `Counttable`?

The former for sure. There's no reason we shouldn't be able to use murmur, rolling, or even two-bit hash with the Counttable. Different hash functions provide different behaviors that are useful in different contexts: two-bit? fast, plus graph traversal! murmur? random dispersal, but slow! rolling?, random dispersal, but (hopefully) fast! In any case, we need to get away from encoding the hash function behavior in the data structure name.

I think the goal here has been to decouple hash functions from storage types completely. @ctb laid some of the important groundwork for that, but there's still more to be done. @luizirber hinted recently that he would be working on this, but his recent lab meeting presentation has been consuming his attention recently and I don't think there's a branch to point at yet. (Let's not forget that he's writing a new JavaScript framework...kind of a big deal!)

**Comment by standage** [comment]
> is it worth exposing these kmer iterators to python land?

The only utility would be for testing, I think. The performance overhead of sending each k-mer to Python land would negate any benefit from having the underlying implementation in C++.

**Comment by luizirber** [comment]
@standage no JS web framework, sir 😸 

I didn't go far into refactoring hashers, I started modifying the HLL code to use the MurmurHash iterator, but then got into a rabbit hole trying to make the iterator a proper C++ iterator to be able to write 
```c++
auto kmers = MurmurKmerHashIterator(inp.c_str(), _ksize);
for(auto hashed: kmers) {
    add(hashed);
```
but then got stuck with details on how to implement the iterator (thanks C++). The branch is [refactor/hasher](https://github.com/dib-lab/khmer/compare/refactor/hasher)

**Comment by betatim** [comment]
What do you think of an interface like:
```
ct_with_murmur = Counttable(..usual args, hash='murmur')
ct_with_murmur = Counttable(..usual args, hash='rolling')
```
with a default value of `'murmur'` (current behaviour). As a user this looks quite nice no?

Next step, how do we make a c++ type that at compile time knows what hash it is using? [Traits](https://erdani.com/publications/traits.html)? https://stackoverflow.com/a/30440624?

**Comment by betatim** [comment]
Probably more like https://stackoverflow.com/questions/6696033/specialize-c-member-function-based-on-class-template-argument

**Comment by ctb** [comment]
interface looks good!

**Comment by camillescott** [comment]
I also have a branch working on a new hashing interface: https://github.com/dib-lab/khmer/blob/feature/swappable_hashfuncs/include/oxli/hashing.hh

Main point of interest is the expected interface for hash functions. I settled on a set of three possible interfaces of a hash function:

* the basic: hashes `string` to `HashIntoType` or `Kmer`
* reversible: supports an `unhash` method
* shiftable: supports converting `HashIntoType` or `kmer` plus the next base to another `HashIntoType` or `Kmer`
* Compositions of the three

I played around with templating some, but decided to use virtual functions for an initial attempt. This lets you overload the `KmerIterator::next` function for a regular `KmerHasher *` and `ShiftingKmerHasher *`. Caveats are that the two-bit hasher would multiply inherit `ShiftingKmerHasher` and `ReversibleKmerHasher`, which some people don't like.

Traits seem like a more modern way to things, which I will likely experiment with.

**Comment by betatim** [comment]
https://github.com/dib-lab/khmer/blob/18799e2fa733aba4a52089c61a42ae7d68131949/khmer/_oxli/graphs.pyx#L860-L868 is an attempt to do something akin to policy based design/traits. Imagine instead of having different languages we'd have different hash functions. One way out is to have a `PyGreeterDE` and a `PyGreeterEN` which we hide behind a `PyGreeter` that users actually use. It instantiates the right `PygreeterLANG` based on the `__init__` argument and then forwards everything to that instance. There must be a better way?

**Comment by betatim** [comment]
I think we are getting somewhere with this. All the tests pass! 🎉

To unique-fy the hashes I used to XOR the forward and reverse values like we do for murmur hash. However that lead to there being only very few unique values (the number of unique hashes of random 8-mers). Not quite sure I understand why. Instead I compare the two values and pick the larger one. However this leads to the following ([script](https://gist.github.com/betatim/2dd610abb66e7741c2ac75912cbabacf):

hashes

The hashes aren't at all uniformly distributed. You get something similar (but slope in the other direction) if you plot the twobit hash values. Not sure what the conclusion is here. Mostly just confused.

**Comment by ctb** [comment]
The slopes are probably because of the comparison - uniqify for twobit takes the lesser of the two, I believe.

**Comment by betatim** [comment]
Agreed that it is because of the comparison. What puzzles me is why XOR doesn't work and why this seems to be kinda innocent given no one noticed so far?

You can make the cyclic hash flat by using `kmer > rev_comp ? h : r` to uniquify. So you do a string comparison to make the decision. A bit awkward though in the `RollingHashIterator` as we don't explicitly construct the reverse complement.

**Comment by ctb** [comment]
this is what we do in sourmash I think

**Comment by ctb** [comment]
interesting and relevant discussion about XOR and combining hashes: https://stackoverflow.com/questions/5889238/why-is-xor-the-default-way-to-combine-hashes

**Comment by betatim** [comment]
hashes

Baam! Adding forward and backward hashes instead of XORing or string compare seems to do the trick. Via this SO answer from Titus' link https://stackoverflow.com/a/27952689.

**Comment by betatim** [comment]
Just for prettyness' sake, using `hash() + 2**64/2` as "pythonhash":
```python
def hash_(kmer):
    rev = khmer.reverse_complement(kmer)
    h1 = hash(kmer) + (2**64)/2
    h2 = hash(rev) + (2**64)/2
    return (h1 + h2) % 2**64
```

So no more negative numbers:

hashes

**Comment by betatim** [comment]
Note to self: this PR currently contains a modified murmurhash (using addition instead of XOR). Discuss before merging.

**Comment by betatim** [comment]
Finally travis is green!

Ready for review! @luizirber, @camillescott, @standage , or @ctb. This is a pretty big PR so I will try and add some comment to places that are interesting. Feel free to comment though.

The [change to the murmur hash'ing](https://github.com/dib-lab/khmer/pull/1792#issuecomment-340015672) has been reverted

File: include/oxli/kmer_hash.hh [hunk 1]
```diff
@@ -409,6 +417,67 @@ public:
...
+    HashIntoType first() {
+        _initialized = true;
+
+        for (char i = 0; i < _ksize; ++i) {
+            fwd_hasher.eat(*(_seq + i));
+            bwd_hasher.eat(_rev[length - _ksize + i]);
+        }
+        index += 1;
+
+        return fwd_hasher.hashvalue + bwd_hasher.hashvalue;
```
[the remark this answers is not in the document]
Comment (betatim) [reply include/oxli/kmer_hash.hh]: Not sure we should feel dirty? We can remove an `if` statement, which should make us rejoice! :)

**Comment by betatim** [comment]
FDR == False discovery rate?

Thanks for the merge!

**Comment by standage** [comment]
> FDR == False discovery rate?

Yep

**Pull request merged**